harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,396 @@
+import abc
+import logging
+import queue
+import threading
+import time
+from datetime import datetime, timezone
+from typing import Any, Dict, List
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        # xxx: pass in train context here to use for metrics reporting API?
+        self._distributed = distributed
+        self._on = False
+
+        self._collector = None
+        self._shipper = None
+
+    # xxx: other potential configs (collection interval, flush interval, max collection time)
+    def on(self, interval: int = 1) -> None:
+        if self._on:
+            logger.warning(f"Profiler is already on.")
+            return
+
+        # Only enable on local chief workers.
+        if self._distributed.local_rank != 0:
+            return
+
+        logger.info(f"Starting system metrics profiling.")
+
+        metrics_queue = queue.Queue()
+        self._collector = _Collector(
+            metrics_queue=metrics_queue,
+            collection_interval=interval,
+        )
+        self._shipper = _Shipper(
+            session=self._session,
+            agent_id=self._agent_id,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            metrics_queue=metrics_queue,
+        )
+
+        self._collector.start()
+        self._shipper.start()
+        self._on = True
+
+    def off(self) -> None:
+        if not self._on:
+            logger.warning(f"Profiler is already off.")
+            return
+        logger.info(f"Stopping system metrics profiling.")
+        self.close()
```

Comment:
pick one; either use `on/off` and let it be a lightswitch that can be flipped back and forth, or let `close()` invalidate the resource and don't let it be reopened.

Practically anything with `.close()` is unusable after you call `.close()`.